clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -566,6 +566,116 @@ class radix_sort {
   uint8_t *_local_memory;
 };
 
+/// Load linear segment items into block format across threads
+/// Helper for Block Load
+enum load_algorithm{
+
+  blocked,
+  striped,
+  warp_striped
+
+};
+
+template <size_t GROUP_WORK_ITEMS,
+          size_t ITEMS_PER_WORK_ITEM,
+          load_algorithm ALGORITHM,
+          typename InputT,
+          typename InputIteratorT>
+class workgroup_load {
+public:
+  
+  __dpct_inline__ void load_blocked(size_t linear_tid, InputIteratorT block_itr,
+                                    InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+  
+  uint32_t workgroup_offset = linear_tid * ITEMS_PER_WORK_ITEM;
+  if (workgroup_offset > GROUP_WORK_ITEMS){
+      uint32_t final_idx = (GROUP_WORK_ITEMS - workgroup_offset);
+      for (uint32_t idx = 0; idx < final_idx ; idx++) {
+        items[idx] = block_itr[workgroup_offset + idx];
+      }
+    }
+    else{
+    #pragma unroll
+      for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+        items[idx] = block_itr[workgroup_offset + idx];
+      }
+    }
+  }
+
+private:
+};
+
+template <size_t GROUP_WORK_ITEMS,
+          size_t ITEMS_PER_WORK_ITEM,
+          load_algorithm ALGORITHM,
+          typename InputT,
+          typename InputIteratorT,
+          typename Item>
+class subgroup_load {
+public:
+  __dpct_inline__ void load_striped(size_t linear_tid, InputIteratorT block_itr,
+                                    InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+
+    if (linear_tid + ((ITEMS_PER_WORK_ITEM - 1)*ITEMS_PER_WORK_ITEM) > GROUP_WORK_ITEMS){
+      uint32_t final_idx = (GROUP_WORK_ITEMS - linear_tid)/ITEMS_PER_WORK_ITEM;
+      for (uint32_t idx = 0; idx < final_idx ; idx++) {
+        items[idx] = block_itr[linear_tid + (idx  * ITEMS_PER_WORK_ITEM)];
+      }
+    }
+    else{
+    #pragma unroll
+      for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+        items[idx] = block_itr[linear_tid + (idx  * GROUP_WORK_ITEMS)];
+      }
+    }
+  }
+
+  
+  __dpct_inline__ void load_subgroup_striped(const Item &item, size_t linear_tid,
```

Comment:
Can you explain the difference between `load_striped` and `load_subgroup_striped`?  I'm not following how these are meant to be used.

This one does a placement `new` operator, and incorporates the subgroup offset of a work_item into the indexing...

An overall description of each APIs semantically in the PR would help me to review this.